pkgs/development/tools/build-managers/meson/setup-hook.sh

```diff
@@ -21,7 +21,9 @@ mesonConfigurePhase() {
         -Dwrap_mode=${mesonWrapMode:-nodownload} \
         $mesonFlags"
 
-    mesonFlags="${crossMesonFlags+$crossMesonFlags }--buildtype=${mesonBuildType:-plain} $mesonFlags"
+    mesonBuildType="${mesonBuildType-plain}"
+    mesonBuildTypeFlag="${mesonBuildType:+--buildtype=${mesonBuildType@Q} }"
+    mesonFlags="${crossMesonFlags+$crossMesonFlags }$mesonBuildTypeFlag$mesonFlags"
```

Comment:
I do not really like the bash shell parameter expansion as I can never remember which combination of symbols does what and it is hard to search for. Using the `if` statement is clearer IMO:

```suggestion
    # By default, we use plain build type. You can override it using mesonBuildType
    # variable or disable passing the flag by setting the variable to empty string.
    mesonBuildType="${mesonBuildType-plain}"
    if [[ -n $mesonBuildType ]]; then
        mesonFlags="--buildtype=${mesonBuildType} $mesonFlags"
    fi
    
    mesonFlags="${crossMesonFlags+$crossMesonFlags }$mesonFlags"
```

The suggestion also removes the ineffectual `@Q` (we should really switch to using an array for all the flags) and adds an explanation comment since the empty string behaviour is easy to miss.

Reply:
Actually, having a special empty value might be confusing as well. I would prefer just using an explicit switch:

```suggestion
    if [[ -z "${dontAddMesonBuildType-}" ]]; then
        mesonFlags="--buildtype=${mesonBuildType:-plain} $mesonFlags"
    fi

    mesonFlags="${crossMesonFlags+$crossMesonFlags }$mesonFlags"
```

Reply:
@jtojnar What makes you think the `@Q` is ineffectual?  I think it is required to make `mesonBuildType` values containing spaces work properly (admittedly it happens there are no valid such, but at least it will make the error behaviour a little less confusing).
Indeed it would ideally be an array, but I'm not keen on changing that interface in this PR about removing a flag.

On the use of parameter expansion features — I think it's quite idiomatic bash, but at any rate I'd rather follow the existing style rather than refactoring (and this refactor uses the `-` parameter expansion form anyway).

Reply:
Ineffectual I referred to https://github.com/NixOS/nixpkgs/pull/82264#issuecomment-603274127 – `$mesonFlags` is [passed to meson](https://github.com/Twey/nixpkgs/blob/1a773610af63351a6bdf91c0427667b831c349f1/pkgs/development/tools/build-managers/meson/setup-hook.sh#L30) so `IFS` splitting occurs and quoting does not help.

Personally, I prefer slightly more verbose `if` expressions since they are much more understandable to people without deep Bash background (and the variable expansion forms are hardly googleable) but I am not categorically opposed to using those Bash features. But I definitely think using an empty string value is too obscure and would vote for `dontAddMesonBuildType` as mentioned [above](https://github.com/NixOS/nixpkgs/pull/82264/files#r397209274).

Reply:
I'm not super keen on the flag because it adds nonsensical cases where the `mesonBuildType` is set but also `dontAddMesonBuildType`.  Ideally I would have had the undefined case not set the build type, and let the user or `mkDerivation` decide when they want to override it, but I think that would change the behaviour of too many packages now, and I agree that the special empty string behaviour is perhaps a little surprising if not documented. 

I've accepted the suggestion: I've a mild bias towards preserving the status quo, but better safe than sorry on the readability aspect. :)